pkg/queue/breaker.go

```diff
@@ -99,25 +128,57 @@ type Semaphore struct {
 	mux      sync.Mutex
 }
 
-// Acquire receives the token from the semaphore, potentially blocking
+// Acquire receives the token from the semaphore, potentially blocking.
 func (s *Semaphore) Acquire() {
 	<-s.queue
 }
 
-// Release releases the token to the queue
-// The operation is potentially blocking when the queue is full
+// Release potentially puts the token back to the queue.
+// If the semaphore capacity was reduced in between and is not yet reflected,
+// we remove the tokens from the rotation instead of returning them back.
 func (s *Semaphore) Release() {
-	s.AddCapacity(1)
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	if s.reducers > 0 {
+		s.capacity -= s.reducers
+		s.reducers--
+	} else {
+		// We want to make sure releasing a token is always non-blocking.
+		select {
+		case s.queue <- s.token:
+		default:
+		}
```

Comment:
Does this swallow errors in case the channel is full?

Reply:
It does indeed - I mimicked the previous logic.
I can add a logger into the semaphore to print it out. Actually the error should never happen (must be careful with saying "never" though). In this case the semaphore's capacity is full - only happens when there are more tokens returned then acquired. I'm not sure if we need to propagate it further up to the Breaker and change its contract. What do you think?

Reply:
I see your point, should we return an error in this case instead of passing a logger in? The caller can log the error then I think.

Reply:
Yeap, fine with me, could save the change in the Semaphore's constructor then.